game/world/managers/objects/units/player/TalentManager.py

```diff
@@ -49,18 +49,37 @@ def send_talent_list(self):
                 if not self.player_mgr.skill_manager.can_ever_use_equipment(spell_item_class, spell_item_subclass_mask):
                     continue
 
+            # Checking magic talents to make sure class can use them.
+            if skill_line_ability.SkillLine == TalentSkillLines.MAGIC_TALENTS:
+                if not TrainerUtils.player_can_learn_magic_talent(training_spell, spell, self.player_mgr):
+                    continue
+
             # Search previous spell.
             preceded_skill_line = DbcDatabaseManager.SkillLineAbilityHolder.skill_line_abilities_get_preceded_by_spell(spell.ID)
             preceded_spell = 0 if not preceded_skill_line else preceded_skill_line.Spell
 
             talent_points_cost = TalentManager.get_talent_cost_by_id(training_spell.playerspell)
-            status = TrainerUtils.get_training_list_spell_status(spell, training_spell.spell, spell.BaseLevel,
+            status = TrainerUtils.get_training_list_spell_status(spell, training_spell, spell.BaseLevel,
                                                                  preceded_spell, self.player_mgr)
 
+            # If the spell before this one exists and is unavailable, don't show this one. (We only want to show the first unavailable spell in a chain.)
+            if preceded_spell != 0:
+                preceded_preceded_skill_line = DbcDatabaseManager.SkillLineAbilityHolder.skill_line_abilities_get_preceded_by_spell(preceded_spell)
+                preceded_preceded_spell = 0 if not preceded_preceded_skill_line else preceded_preceded_skill_line.Spell
+
+                preceded_spell_entry: Optional[Spell] = DbcDatabaseManager.SpellHolder.spell_get_by_id(preceded_spell)
+                preceded_trainer_spell_id = WorldDatabaseManager.TrainerSpellHolder.trainer_spell_id_get_from_player_spell_id(WorldDatabaseManager.TrainerSpellHolder.TRAINER_TEMPLATE_TALENT_ID, preceded_spell)
+                preceded_trainer_spell = WorldDatabaseManager.TrainerSpellHolder.trainer_spell_entry_get_by_trainer_and_spell(WorldDatabaseManager.TrainerSpellHolder.TRAINER_TEMPLATE_TALENT_ID, preceded_trainer_spell_id)
+                preceded_status = TrainerUtils.get_training_list_spell_status(preceded_spell_entry, preceded_trainer_spell, preceded_spell_entry.BaseLevel, 
+                                                                              preceded_preceded_spell, self.player_mgr)
+                                                                              
+                if preceded_status == TrainerServices.TRAINER_SERVICE_UNAVAILABLE:
+                    status = TrainerServices.TRAINER_SERVICE_NOT_SHOWN
+
```

Comment:
Would it be enough here to check if the player already knows the preceding (preceding) spell? Using `get_training_list_spell_status` adds level and skill checks, but afaik these don't vary between talent ranks.
e.g.
```py
            if preceded_spell:
                preceded_preceded_skill_line = DbcDatabaseManager.SkillLineAbilityHolder.skill_line_abilities_get_preceded_by_spell(preceded_spell)
                preceded_preceded_spell = 0 if not preceded_preceded_skill_line else preceded_preceded_skill_line.Spell
                if preceded_preceded_spell and preceded_preceded_spell not in self.player_mgr.spell_manager.spells:
                    status = TrainerServices.TRAINER_SERVICE_NOT_SHOWN

Reply:
I tried something similar along with this snippet just now, however unfortunately this only works up to Rank 2. In this case it leads to Ranks 1 and 2 being shown in unavailable, instead of just Rank 1. (ex. Sword Power (Rank 1), Sword Power (Rank 2) are both shown in unavailable if the player does not have the Swords skill.)

Reply:
I see; since the above snippet doesn't consider skill requirements, we assume there's no other requirements for Rank 1. The below snippet would fix the case you mentioned with skill requirements, but now reconsidering I think the existing code is more future-proof. This snippet again assumes no other restrictions for talent learning (ie. skill level, level requirement) that could be added later to `get_training_list_spell_status`.

```py
            if preceded_spell:
                preceded_preceded_skill_line = DbcDatabaseManager.SkillLineAbilityHolder.skill_line_abilities_get_preceded_by_spell(preceded_spell)
                preceded_preceded_spell = 0 if not preceded_preceded_skill_line else preceded_preceded_skill_line.Spell
                if preceded_preceded_spell and preceded_preceded_spell not in self.player_mgr.spell_manager.spells or \
                        training_spell.reqskill not in self.player_mgr.skill_manager.skills:
                    status = TrainerServices.TRAINER_SERVICE_NOT_SHOWN